File: MATLAB/02631_IntrProg/week006/computeLanguageError.m
function se = computeLanguageError(freq)
% takes as input a vector of frequencies of 
% occurrences of letters in a text. 
% read the file letter_frequencies.csv, 
% compute the squared error for each language, 
% and return a vector of squared errors for the fifteen languages.

% INPUT 
% freq  :   A vector of size 26 containing the 
%           frequency of the letters a?z in a text.
% OUTPUT
% se    :   A vector of length 15 containing the squared error 
%           between the input vector and each of the 
%           15 languages in the CSV file

lFreq = readtable('letter_frequencies.csv');
[~,c] = size(lFreq);
se = zeros(1,c-1);
for i = 2:c
    se(i-1) = sum((freq'-lFreq{:,i}).^2);
end
end

File: python/02631_IntrProg/week006/computeLanguageError.py
import numpy as np
import pandas as pd
 
def computeLanguageError(freq):
    # Computes the squared error between the letter frequencies given
    # as input and the letter frequencies of 15 languages stored in the
    # file letter_frequencies.csv
 
    # Open the file letter_frequencies.
    data = pd.read_csv("letter_frequencies.csv")
 
    # Extract a matrix containing the letter frequencies for the 15
    # languages and transpose it to yield a size 15 x 26 matrix
    languageFrequencies = np.array(data.iloc[0:26, 1:16]).T
 
    # Compute the squared errors. Subtracting a vector from a matrix will
    # "recycle" the vector, so it is subtracted from each row.
    E = np.sum((languageFrequencies - freq)**2, axis=1)
    return E